drv/cosmo-hf/src/main.rs

```diff
@@ -0,0 +1,294 @@
+    SectorEraseBusy,
+    WriteBusy,
+
+    HashUpdateError(drv_hash_api::HashError),
+    HashFinalizeError(drv_hash_api::HashError),
+}
+
+counted_ringbuf!(Trace, 32, Trace::None);
+
+/// Size in bytes of a single page of data (i.e., the max length of slice we
+/// accept for `page_program()` and `read_memory()`).
+///
+/// This value is really a property of the flash we're talking to and not this
+/// driver, but it's correct for all our current parts. If that changes, this
+/// will need to change to something more flexible.
+pub const PAGE_SIZE_BYTES: usize = 256;
+
+/// Size in bytes of a single sector of data (i.e., the size of the data erased
+/// by a call to `sector_erase()`).
+///
+/// This value is really a property of the flash we're talking to and not this
+/// driver, but it's correct for all our current parts. If that changes, this
+/// will need to change to something more flexible.
+///
+/// **Note:** the datasheet refers to a "sector" as a 4K block, but also
+/// supports 64K block erases, so we call the latter a sector to match the
+/// behavior of the Gimlet host flash driver.
+pub const SECTOR_SIZE_BYTES: usize = 65_536;
+
+#[export_name = "main"]
+fn main() -> ! {
+    // Wait for the FMC to be configured
+    userlib::hl::sleep_for(1000);
+
+    let id = unsafe { reg::BASE.read_volatile() };
+    assert_eq!(id, 0x1de);
+
+    // Fire up a server.
+    let mut server = ServerImpl;
+    server.flash_set_quad_enable();
+    let mut buffer = [0; hf::idl::INCOMING_SIZE];
+    loop {
+        idol_runtime::dispatch(&mut buffer, &mut server);
+    }
+}
+
+struct ServerImpl;
+
+#[allow(unused)]
+mod reg {
+    pub const BASE: *mut u32 = 0x60000000 as *mut _;
+
+    pub const NOR: *mut u32 = BASE.wrapping_add(0x40);
+    pub const SPICR: *mut u32 = NOR.wrapping_add(0x0);
+    pub const SPISR: *mut u32 = NOR.wrapping_add(0x1);
+    pub const ADDR: *mut u32 = NOR.wrapping_add(0x2);
+    pub const DUMMY_CYCLES: *mut u32 = NOR.wrapping_add(0x3);
+    pub const DATA_BYTES: *mut u32 = NOR.wrapping_add(0x4);
+    pub const INSTR: *mut u32 = NOR.wrapping_add(0x5);
+    pub const TX_FIFO: *mut u32 = NOR.wrapping_add(0x6);
+    pub const RX_FIFO: *mut u32 = NOR.wrapping_add(0x7);
+}
+
+#[allow(unused)]
+mod instr {
+    pub const PAGE_PROGRAM: u32 = 0x02;
+    pub const READ: u32 = 0x03;
+    pub const READ_STATUS_1: u32 = 0x05;
+    pub const READ_STATUS_2: u32 = 0x35;
+    pub const READ_STATUS_3: u32 = 0x15;
+    pub const WRITE_STATUS_2: u32 = 0x31;
+    pub const WRITE_ENABLE: u32 = 0x06;
+    pub const FAST_READ_QUAD: u32 = 0x6b;
+    pub const FAST_READ_QUAD_OUTPUT_4B: u32 = 0x6c;
+    pub const SECTOR_ERASE: u32 = 0x20;
+    pub const READ_JEDEC_ID: u32 = 0x9f;
+    pub const BLOCK_ERASE_64KB: u32 = 0xd8;
+    pub const BLOCK_ERASE_64KB_4B: u32 = 0xdc;
+    pub const QUAD_INPUT_PAGE_PROGRAM: u32 = 0x32;
+    pub const QUAD_INPUT_PAGE_PROGRAM_4B: u32 = 0x34;
+}
+
+impl ServerImpl {
+    fn flash_read_id(&self) -> [u8; 20] {
```

Comment:
**praise:** For an outsider like me, it's really cool seeing this lowest-of-the-low level register work happening here.

**suggestion:** A mountain of comments wouldn't maybe go amiss here and would be a terrific pleasure to read :)